### PR TITLE
[fsdp] Replace acc_grad hooking with register_post_accumulate_grad_hook on flat_param

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115410
* #115497
* __->__ #112184
* #115315

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

### CODE REVIEW

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -1151,15 +1151,10 @@ def _finalize_params(
...
-        _p_assert(
-            post_backward_hook_state_len == expected_post_backward_hook_state_len,
-            f"Invalid: ``_post_backward_hook_state``: {flat_param._post_backward_hook_state}",
-        )
-        flat_param._post_backward_hook_state[-1].remove()
-        delattr(flat_param, "_post_backward_hook_state")
+    if hasattr(flat_param, "_post_backward_hook_handle"):
+        pbhs_handle = flat_param._post_backward_hook_handle
+        pbhs_handle.remove()
+        delattr(flat_param, "_post_backward_hook_handle")
```
Comment (albanD) [inline torch/distributed/fsdp/_runtime_utils.py:1157]: Very much style preference, feel free to ignore
```suggestion
        flat_param._post_backward_hook_handle.remove()
        del flat_param._post_backward_hook_handle
```

File: torch/distributed/fsdp/_runtime_utils.py [hunk 2]
```diff
@@ -1422,12 +1417,9 @@ def _register_post_backward_hook(
...
         "register the post-backward hook",
     )
-    acc_grad = temp_flat_param.grad_fn.next_functions[0][0]  # type: ignore[union-attr]
-    assert acc_grad is not None
-    hook_handle = acc_grad.register_hook(
-        functools.partial(_post_backward_hook, state, handle)
-    )
-    flat_param._post_backward_hook_state = (acc_grad, hook_handle)  # type: ignore[attr-defined]
+    hook = functools.partial(_post_backward_hook, state, handle)
+    hook_handle = temp_flat_param.register_post_accumulate_grad_hook(hook)
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:1421]: I think?
```suggestion
    hook_handle = flat_param.register_post_accumulate_grad_hook(hook)
```

Comment (albanD) [inline torch/distributed/fsdp/_runtime_utils.py:1421]:
```suggestion
    hook_handle = flat_param.register_post_accumulate_grad_hook(hook)
```

And you can remove the code creating temp_flat_param above.
This is also the cause of CI failures saying you can only register this hook on leafs.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 3]
```diff
@@ -1467,7 +1459,7 @@ def _register_post_backward_reshard_only_hook(
     hook_handle = register_multi_grad_hook(
         inp_tensors, functools.partial(_post_backward_reshard, state, handle)
     )
-    flat_param._post_backward_hook_state = (hook_handle,)  # type: ignore[attr-defined, assignment]
+    handle.flat_param._post_backward_hook_handle = hook_handle # type: ignore[attr-defined, assignment]
```
Comment (fegin) [inline torch/distributed/fsdp/_runtime_utils.py:1462]:
```suggestion
    flat_param._post_backward_hook_handle = hook_handle # type: ignore[attr-defined, assignment]
```
nit: this should be consistent with the above code.

**Review by albanD — APPROVED** [review]
Nice cleanup :)

**Comment by voznesenskym** [comment]
hmm one little failure, looking

**Comment by voznesenskym** [comment]
> I guess the one difference between these two that I see is that the old code manually attaches the hook to the current AccGrad node while the new code uses the proper logic of storing the hook on the Tensor and having the AccGrad node read it from there when it is needed.
> 
> In a non-.data= world, these two should really do the same thing. And even if set_() is used, both the old and new AccGrad nodes will point to the same variable and so both will have the hook.

So, what do you suggest we do to reconcile it?

**Comment by albanD** [comment]
> So, what do you suggest we do to reconcile it?

I'm not sure what to do tbh: you go from a "wrong" state to another "wrong" state.
Hard to say which one is less wrong though.

The easy answer would be to just trace the per-parameter which seems to be doing the right thing there.

**Comment by voznesenskym** [comment]
> > So, what do you suggest we do to reconcile it?
> 
> I'm not sure what to do tbh: you go from a "wrong" state to another "wrong" state. Hard to say which one is less wrong though.
> 
> The easy answer would be to just trace the per-parameter which seems to be doing the right thing there.

I might just do the hook refactor w/ torch.compile only, leaving eager as is, let's talk about this? wrong->wrong is one framing, but wrong(not compiling) -> wrong(compiling) is a win in my (sad) book.

**Pull request closed without merging**